xdsl/transforms/print_to_println.py

```diff
@@ -0,0 +1,180 @@
+    """
+    val = re.sub(r"(\s+|\.)", "_", val)
+    val = re.sub(r"[^A-Za-z0-9_]+", "", val).rstrip("_")
+    return val
+
+
+def _key_from_str(val: str) -> str:
+    """
+    Generate a symbol name form any given string.
+
+    Takes the first ten letters of the string plus it's sha1 hash to create a
+    (pretty much) globally unique symbol name.
+    """
+    h = hashlib.new("sha1")
+    h.update(val.encode())
+    return f"{legalize_str(val[:10])}_{h.hexdigest()}"
+
+
+def _format_string_spec_from_print_op(op: print.PrintLnOp) -> Iterable[str | SSAValue]:
+    """
+    Translates the op:
+    print.println "val = {}, val2 = {}", %1 : i32, %2 : f32
+
+    into this sequence:
+    ["val = ", %1, ", val2 = ", %2]
+
+    Empty string parts are omitted.
+    """
+    format_str = op.format_str.data.split("{}")
+    args = iter(op.format_vals)
+
+    for part in format_str[:-1]:
+        if part != "":
+            yield part
+        yield next(args)
+    if format_str[-1] != "":
+        yield format_str[-1]
+
+
+def _format_str_for_typ(t: Attribute):
+    match t:
+        case builtin.f64:
+            return "%f"
+        case builtin.i32:
+            return "%i"
+        case builtin.i64:
+            return "%li"
+        case _:
+            raise ValueError(f"Cannot find printf code for {t}")
+
+
+class PrintlnOpToPrintfCall(RewritePattern):
+    collected_global_symbs: dict[str, llvm.GlobalOp]
+
+    def __init__(self):
+        self.collected_global_symbs = dict()
+
+    def _construct_global(self, val: str):
+        """
+        Constructs an llvm.global operation containing the string.
+        """
+        data = val.encode() + b"\x00"
+
+        t_type = builtin.TensorType.from_type_and_list(i8, [len(data)])
+
+        return llvm.GlobalOp.get(
+            llvm.LLVMArrayType.from_size_and_type(len(data), i8),
+            _key_from_str(val),
+            constant=True,
+            linkage="internal",
+            value=builtin.DenseIntOrFPElementsAttr.from_list(t_type, data),
+        )
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: print.PrintLnOp, rewriter: PatternRewriter, /):
+        format_str = ""
+        args: list[SSAValue] = []
+        casts: list[Operation] = []
+        # make sure all arguments are in the format libc expects them to be
+        # e.g. floats must be promoted to double before calling
+        for part in _format_string_spec_from_print_op(op):
+            if isinstance(part, str):
+                format_str += part
+            elif isinstance(part.typ, builtin.IndexType):
+                # index must be cast to fixed bitwidth (I think) before printing
+                casts.append(new_val := arith.IndexCastOp.get(part, builtin.i64))
+                args.append(new_val.result)
+                format_str += "%li"
+            elif part.typ == builtin.f32:
+                # f32 must be promoted to f64 before printing
```

Comment:
I wonder whether this should be done as a separate rewrite, seems kind of dodgy to me. We could borrow python's behaviour of there being some default print format and a trait for conversion to printf-supported types.

Reply:
hmmm, interesting point. I think this should still happen here though, because the type printing code shouldn't be bothered with the intricacies of the `printf` function?

Also, we plan on supporting both `printf` based lowering and a `putc` based one, so we need to be careful how we generalize here...

Reply:
@superlopuh I think what you are saying is that you would like to lower to a printfOp first and then from printfOp to putchar?
Because then it makes more sense to me to change the current println op?
Could you be more precise with what you mean with "default print format?"

Reply:
I mean in general it would be good to know what the semantics are of what is printable and how. I'd rather just fail to print certain types than have implicit conversions for the time being, hopefully clients can work around in the near future.

Reply:
`PrintableType` interface is on the horizon, I want to be able to print memrefs, etc!